### ISSUE TITLE
DJI Mini video stream issues

### ISSUE THREAD

**Issue by robbiedood** [RosettaDrone/rosettadrone#27]
Dear rosettadrone friends,

In QGroundControl, we are able to take off, landing, got summary of drone parameters.
However, the video panel is always showing "waiting for video", even though we've set Video Source to UDP Video Stream (tried both H.264 and H.265), and set UDP Port to 5600. 

We tried both internal QGC in the same android device with rossettadrone and external QGC in a separate Ubuntu 18.04 desktop.

We were wondering if someone got the same issue with us. 
Our drone is DJI Mavic 2 Zoom & Mavic Mini, Mobile Pixel 2 / 3XL / 5 in Android 11, tried both master and VIDEO_WORK branch.

Hope to hear your expertise.

Best,
-- Luke

**Comment by The1only** [comment]
That should work perfect, check the firewall settings and try using netcat to trace the udp stream. I uses that setup every day, with very low latency.

**Comment by The1only** [comment]
Ps: there are some issues with some versions of gradle, do NOT update anything untill you can make working code, then update gradually.

**Comment by lukelu520** [comment]
@The1only , thanks for your support. Which gradle version are you using for the best performance ? We can downgrade to that working version.

For example, we downgraded to Android Gradle Plugin 3.5.3 & Gradle 6.2.2, situation improves, however, the video stream is fragile as shown in below attached image. The fragile effect happens when an object is moving (if everything is static, then it's fine).

![rosetta_drone_video_fragile](https://user-images.githubusercontent.com/36803293/99158760-310a4b80-268b-11eb-9e5e-8abf5ffbd386.jpg)


Hope to hear your expertise. Then we could work on the code to fix the issue.

Thanks,
-- Luke

**Comment by The1only** [comment]
I use 3.5.3 and 6.3, the video with the Mavic 2 zoom is supose to be perfect.

**Comment by lukelu520** [comment]
Update: We disabled "Enable Video feed" in setting, and video becomes perfect. It's interesting, Do you know the reason ?

Best,
-- Luke

**Comment by The1only** [comment]
Hmm have not tested that in a while, disable the secondary or primary video feed?

Thank you, I will investigate that.

**Comment by lukelu520** [comment]
Thanks for your prompt reply! We disabled all of them in Video of Setting.  (we expected video would disappear, but turns out it works perfectly.) 

Update: we now use Android Gradle Plugin 3.5.3 & Gradle 6.3 as your suggestion.  The video fragile issue could be resolved by turn off and turn on "Enable video feed" switcher. Now our "Enable video feed" needs to be on. 

It's kind of doing reset ?

Best,
-- Luke

**Comment by The1only** [comment]
I added variables to the config file, maybe the system was still using an old config file with the new settings, and when flipping a new file was written and the problem fixed....  must remember that for later updates. 

But hey... you got video!

**Comment by The1only** [comment]
Just passed a majore computer vission AI project, flying live with a DJI Mavic and a CNN AI running on an eGPU on the QGC computer. The drone used Rosettadrone video and telemetry for control doing complex movements... happy days...

**Comment by lukelu520** [comment]
@The1only , thanks for your support. After making mavic 2 zoom video working, we are now working on mavic mini.
We added mavic mini as a new case in validateTranscodingMethod and received some juices, however, the video is noisy signal. (seems like a bad encoding/decoding) 
Could you tell us which config file (or the portion of code) we could check in order to make mavic mini working ?

**Comment by The1only** [comment]
First try rebooting EVERYTHING, and try fresh. If that does not help, then it gets complicated, we will have to work our the NAILs and framing... it can take a few weeks to work out.  I do not have a mini my self.

Ps: the old encoding requires more juce and banwith, make sure you are on gb net, and got a powerfull android device.

**Comment by The1only** [comment]
We might at some point have to reintroduce gStreamer to be able to set transfer quality again

**Comment by lukelu520** [comment]
@The1only , thanks for the advice. Bad news is rebooting doesn't help, MINI's video stream shows bad encoding even on the screen of our Android device. (Note that: by using DJI's Fly App, MINI's video stream is well )

Based on the source code, MINI would use 
`NativeHelper.getInstance().parse(videoBuffer, size, mode: 0);` and Send raw H264 to the FFMPEG parser. 

Then it goes to dji_video_jin.c, we then lose the track of the detail of FFMPEG parser; were wondering where we should focus on tuning parameters for MINI H264 encoding. 

Also, since Mini just become compatible by the latest release of DJI mobile sdk 4.13.1, there might be a new sdk method to do the codec. such as [DJIVideoStreamDecoder.java
](https://developer.dji.com/mobile-sdk/documentation/sample-code/index.html?awc=7327_1605405664_55c3ac9e5aacb625c56e42c96f23a6cd&pbc=awin2017)


Best,
-- Luke

In the same time, we are checking DJI mobile sdk references, figuring out the proper interpretation for DJI Mavic MINI raw video format. If you have any reference on ffmpeg parser / encoding / decoding, feel free to let us know.

References:
[DJI's Android Video Stream Decoding ](https://developer.dji.com/mobile-sdk/documentation/sample-code/index.html?awc=7327_1605405664_55c3ac9e5aacb625c56e42c96f23a6cd&pbc=awin2017)
https://github.com/dji-sdk/Mobile-SDK-Android/issues/592

**Comment by The1only** [comment]
Aha if there is bad video on the android device that is good news, as the SDK IS suppose to support that. Only the exporting is our responsebility.

Now take the video example made by DJI from git hub, see if that works, it should. Then from that start to compare.

**Comment by The1only** [comment]
I bought a Mavic Mini today, and I see the same problem. On the Android device I have perfect video, but on QGC I got crap... I also got an Mavic AIR I will retest this drone, I hade belived they are similar.... 

Update: Retested Mavic Air, it still works, as do Inspire 1  that also uses the old transcoding. So Mavic Mini is a problem...

**Comment by lukelu520** [comment]
It's glad to hear you got a Mini and it works on your android device. (my android device got crap as shown below)

By reviewing the ffmpeg part (i.e. when mode=0) of rosetta's dji_video_jni.c, we can't find a significant difference from DJI's Android Video Stream Decoding Sample (AVSD)'s [dji_video_jni.c ](https://github.com/DJI-Mobile-SDK-Tutorials/Android-VideoStreamDecodingSample/blob/master/android-videostreamdecodingsample/jni/dji_video_jni.c) 

However, we CAN NOT find few .c / .h files in rosetta's folder, e.g. cmdutils.c, cmdutils_common_opts.h, ffmpeg_filter.c, ffmpeg_mod.c, ffmpeg_opt.c, and Application64.mk.

In the same time, we built and installed [DJI's AVSD Sample](https://github.com/DJI-Mobile-SDK-Tutorials/Android-VideoStreamDecodingSample); we found it's working -- the streaming works well (as attached in below) and it's able to output YUV data from Mini.

Maybe we could utilize DJI's AVSD streaming module, integrate it into rosetta drone ?

Best,
-- Luke  

rosettae drone's screen shot on android device (google Pixel 3XL):
![Screenshot (Nov 15, 2020 3_40_25 PM)](https://user-images.githubusercontent.com/36803293/99200266-26b68300-2759-11eb-9ba7-a4b161f80743.png)


DJI AndroidVideoStreamDecodingSample's screenshot on android device (google Pixel 3XL):
![Screenshot (Nov 15, 2020 4_00_21 PM)](https://user-images.githubusercontent.com/36803293/99200788-47cca300-275c-11eb-9851-c557d26c7188.png)

**Comment by The1only** [comment]
Hmm, on my system I got perfect video on the Android device. Remember that the ffmpeg stuff is used for the encoding, while the android device screen is using the decoding from the sdk. If the device video is crap, then there is no need to look at the C code.

**Comment by lukelu520** [comment]
I remember your Android device is Samsung Note20 phone. What's the android version on the phone ?
We'll see if we could get the same phone and system version with you. 
Since it's super weird that Pixel 2 / Pixel 3 even can not display the video on its screen well (like you said, it doesn't involve ffmpeg codec).

**Comment by lukelu520** [comment]
@The1only , thanks for your information. 

Update: We got perfect video both on the Android device and on QGC now.
Again, turn off "Enable Video Feed" and return to main screen, and come back setting and turn on "Enable Video Feed" solve the problem.

However, we got crap video when using gst-launch-1.0

Hope to hear your expertise on gst-launch-1.0 in order to make it work as well.

**Comment by The1only** [comment]
The on and off problem seems to be a sync issue, I hade the same and rebooted the drone and it got fine. But you're method is better. After we get the video to work on QGC then I will look into this.

**Comment by The1only** [comment]
What QGC version do you use that works???

Interesting I am still using QGC 3.6 as the 4.x gives me latency problems. What version og gst 1.x do you use?

I to need gst_launch to work for my project

**Comment by lukelu520** [comment]
We are using QGC 4.0.11 and gst-launch-1.0 version 1.14.5 (GStreamer 1.14.5). We are surprised that QGC works but gst doesn't work on Mini (we thought QGC also uses gstreamer to process the video).  This discrepancy only happens on Mini. Maybe our gst pipeline is not correctly set for Mini, we are checking this [QGC gstreamer page](https://github.com/mavlink/qgroundcontrol/blob/master/src/VideoReceiver/README.md) for a solution.

Thanks for pointing sync issue out. So far we change the default setting of mExternalVideoOut and pref_enable_video to false.
But we were wondering if there is a better way to fix the sync issue. Maybe auto-set mExternalVideoOut to false and then re-set to true when TextureView is set or wait 1 sec after onResume is called ?

Hope to hear your expertise.

**Comment by lukelu520** [comment]
Update: problem solved, we fixed gstreamer pipeline issue; we could live stream MINI's video on Android device / QGC / gstreamer now.

Then there is only issue remain -- the sync issue, which requires us disable "Enable Video Feed" every time when finishing the use of the app.

We'll dig into the problem to see we could make a contribution.

Best,
-- Luke

**Comment by The1only** [comment]
Do tell, what and how did you fix the gStreamer pipeline?

**Comment by lukelu520** [comment]
@The1only , sure, regards to the gstreamer issue, please check the [PR#9180](https://github.com/mavlink/qgroundcontrol/pull/9180) we submitted to mavlink/qgroundcontrol.

Regards to the video sync issue on rosetta drone, let us know if any suggestion or guide to resolve it.
Hope to find a better way to fix it !

Best,
-- Luke

**Comment by The1only** [comment]
So PR#9180 actually made a difference in this case?

As an experiment we can try always start with video off, and then set the selected vodep state after alll system is up.

**Comment by lukelu520** [comment]
No, this is why I didn't post the PR here until you asked for that.

We recently find MINI's h264parse and rtpsocket might have k-frame or header error. This minor issue propagates and impact later stages of application. But it is out of the topic. Maybe I should close the thread.

Thanks for your active support!

--Luke

**Comment by kripper** [comment]
Hi @lukelu520,

I want to fix the problem with the MINI's RTP video stream.
I'm experiencing the same problem as you with my MINI SE.
Does it really fix the problem to disable and re-enable the video feed setting? I'm doing it, but it doesn't fix for me.
What command line are you using for gstreamer?

I found out that 'isKeyFrame' is always `false` in `onDataRecv()`. Maybe that is causing the sync problem.
I also noticed this method receives buffers of only 10.000 bytes. I'm not sure if this is enough for a 1280x720 stream.
